lib/logstash/outputs/google_pubsub.rb

```diff
@@ -0,0 +1,145 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+require 'logstash/outputs/base'
+require 'logstash/namespace'
+require 'logstash/outputs/pubsub/client'
+
+#
+# === Summary
+#
+# A LogStash plugin to upload log events to https://cloud.google.com/pubsub/[Google Cloud Pubsub].
+# Events are batched and uploaded in the background for the sake of efficiency.
+# Message payloads are serialized JSON representations of the events.
+#
+# === Environment Configuration
+#
+# To use this plugin, you must create a
+# https://developers.google.com/storage/docs/authentication#service_accounts[service account]
+# and grant it the publish permission on a topic.
+# The topic __must__ exist before you run the plugin.
+#
+# === Usage
+#
+# There are two major ways to authenticate to Google Cloud Pubsub:
+#
+# * autodiscovered
+#   https://cloud.google.com/docs/authentication/production[Application Default Credentials]
+# * a GCP Service Account JSON file
+#
+# Example configuration:
+#
+# [source,ruby]
+# ------------------------------------------------------------------------------
+# output {
+#   google_pubsub {
+#     # Required attributes
+#     project_id => "my_project"                                      (required)
+#     topic => "my_topic"                                             (required)
+#
+#     # Optional if you're using app default credentials
+#     json_key_file => "service_account_key.json"                     (optional)
+#
+#     # Options for configuring the upload
+#     message_count_threshold => 100                                  (optional)
+#     delay_threshold_secs => 5                                       (optional)
+#     request_byte_threshold => 4096                                  (optional)
+#     attributes => {"key" => "value"}                                (optional)
+#   }
+# }
+# ------------------------------------------------------------------------------
+#
+class LogStash::Outputs::GooglePubsub < LogStash::Outputs::Base
+  config_name 'google_pubsub'
+
+  concurrency :shared
+
+  # Google Cloud Project ID (name, not number)
+  config :project_id, validate: :string, required: true
+
+  # Google Cloud Pub/Sub Topic. You must create the topic manually.
+  config :topic, validate: :string, required: true
+
+  # If logstash is running within Google Compute Engine, the plugin will use
+  # GCE's Application Default Credentials. Outside of GCE, you will need to
+  # specify a Service Account JSON key file.
+  config :json_key_file, validate: :path, required: false
+
+  # Send the batch once this delay has passed, from the time the first message
+  # is queued. (> 0, default: 5)
+  config :delay_threshold_secs, validate: :number, required: true, default: 5
```

Comment:
While having a default value here makes the config item optional, having `required: true` here may cause confusion with the comment above stating the field is optional